### ISSUE TITLE
KEB: Allow specifing private IP address range when provisiong Kyma runtime

### ISSUE THREAD

**Issue by PK85** [kyma-project/control-plane#2942]
**Description**
When that will be supported by the Provisioner, the KEB needs to expose such an option for users. This is only exposed when a user creates a cluster. Updates are not allowed.

**AC**
- [x] 1) Use API from provisioner (ticket: https://github.com/kyma-project/control-plane/issues/2946)
  - [x] a) contact frogs and leave your feedback about their API design, check how new API validates new inputs
  - [x] b) PB suggested that one additional field will be required to narrow down ranges, check with frogs
 PR: https://github.com/kyma-project/control-plane/pull/3020
- [x] 2) Adjust schema with new input parameter for SKR creation (only creation input parameters)
  - affected plans:
    - aws, OpenStack, free, gcp, azure, azure_lite
  - all new params under `networking` section (same like `oidc` parameters)
- [x] 3) check how we can validate input in the schema itself
- [x] 4) In case of wrong input params values, return failure for this SKR with proper message
- [x] 5) Add new skr-test scenario with `networking` params in use

**Links**
- https://gardener.cloud/docs/gardener/usage/shoot_networking/

**Comment by Disper** [comment]
Hiho @kyma-project/gopher,
please see [this comment](https://github.com/kyma-project/control-plane/issues/2946#issuecomment-1674290867) regarding API proposal. Looking forward to your feedback.

Regarding the part:
>check how new API validates new inputs

We're relying on the gardener validation that is implemented [here](https://github.com/gardener/gardener/blob/4f853e308500f57cad5c881227307fbc865736d5/pkg/utils/validation/cidr/cidr.go).

<details>
  <summary>Example validation - must be valid canonical CIDR </summary>
  
  ``` 
  "message": "Failed to start provisioning, error creating Shoot for fbb1effd-99d9-4dd3-8bf6-89b639b4ecf4 cluster: %!!(MISSING)!(MISSING)s(MISSING), Shoot.core.gardener.cloud \"md-aws-cidr\" is invalid: [spec.networking.pods: Invalid value: \"10.250.0.1/19\": must be valid canonical CIDR, spec.networking.services: Invalid value: \"10.250.0.2/19\": must be valid canonical CIDR]",
  ```
</details>

<details>
  <summary>Example validation - must not overlap </summary>

   ``` 
"message": "Failed to start provisioning, error creating Shoot for 37c5736a-07f7-479a-93ad-a8ba57d70066 cluster: %!!(MISSING)!(MISSING)s(MISSING), admission webhook \"validation.aws.provider.extensions.gardener.cloud\" denied the request: [networking.pods: Invalid value: \"10.250.0.0/19\": must not overlap with \"networks.vpc.cidr\" (\"10.250.0.0/16\"), networking.services: Invalid value: \"10.250.0.0/19\": must not overlap with \"networks.vpc.cidr\" (\"10.250.0.0/16\"), networks.zones[0].workers: Invalid value: \"10.250.0.0/19\": must not overlap with \"networking.pods\" (\"10.250.0.0/19\"), networks.zones[0].workers: Invalid value: \"10.250.0.0/19\": must not overlap with \"networking.services\" (\"10.250.0.0/19\")]",
  ```

</details>

**Comment by piotrmiskiewicz** [comment]
hi @pbochynski , could you elaborate 1.b

**Comment by pbochynski** [comment]
There are 2 settings in shoot:
  networking:
    nodes: 10.180.0.0/16

and
  provider:
    infrastructureConfig:
      apiVersion: gcp.provider.extensions.gardener.cloud/v1alpha1
      kind: InfrastructureConfig
      networks:
        workers: 10.180.0.0/16

It is not required tohave such big  IP ranges. 10.180.0.0/25 should be sufficient. It will allow us to create VPC peering easier.

**Comment by piotrmiskiewicz** [comment]
This is how IP ranges are set for AWS,GCP,AZure:


# GCP

```yaml
 networking:
   pods: 100.64.0.0/12
   nodes: 10.250.0.0/19
   services: 100.104.0.0/13

   provider:
     type: gcp
     controlPlaneConfig:
       kind: ControlPlaneConfig
       apiVersion: gcp.provider.extensions.gardener.cloud/v1alpha1
       zone: us-central1-a
     infrastructureConfig:
       kind: InfrastructureConfig
       apiVersion: gcp.provider.extensions.gardener.cloud/v1alpha1
       networks:
         worker: 10.250.0.0/19
         workers: 10.250.0.0/19
```

# AWS

```yaml
  networking:
    pods: 100.64.0.0/12
    nodes: 10.250.0.0/16
    services: 100.104.0.0/13
    
  provider:
    type: aws
    controlPlaneConfig:
      apiVersion: aws.provider.extensions.gardener.cloud/v1alpha1
      cloudControllerManager:
        useCustomRouteController: true
      kind: ControlPlaneConfig
      storage:
        managedDefaultClass: true
    infrastructureConfig:
      kind: InfrastructureConfig
      apiVersion: aws.provider.extensions.gardener.cloud/v1alpha1
      networks:
        vpc:
          cidr: 10.250.0.0/16
        zones:
          - name: us-east-1d
            internal: 10.250.48.0/20
            public: 10.250.32.0/20
            workers: 10.250.0.0/19
          - name: us-east-1a
            internal: 10.250.112.0/20
            public: 10.250.96.0/20
            workers: 10.250.64.0/19
          - name: us-east-1f
            internal: 10.250.176.0/20
            public: 10.250.160.0/20
            workers: 10.250.128.0/19
```

# Azure

```yaml
  networking:
    type: calico
    pods: 100.64.0.0/12
    nodes: 10.250.0.0/16
    services: 100.104.0.0/13

  provider:
    type: azure
    controlPlaneConfig:
      kind: ControlPlaneConfig
      apiVersion: azure.provider.extensions.gardener.cloud/v1alpha1
    infrastructureConfig:
      kind: InfrastructureConfig
      apiVersion: azure.provider.extensions.gardener.cloud/v1alpha1
      networks:
        vnet:
          cidr: 10.250.0.0/16
        zones:
          - name: 3
            cidr: 10.250.0.0/19
            natGateway:
              enabled: true
              idleConnectionTimeoutMinutes: 4
          - name: 2
            cidr: 10.250.32.0/19
            natGateway:
              enabled: true
              idleConnectionTimeoutMinutes: 4
          - name: 1
            cidr: 10.250.64.0/19
            natGateway:
              enabled: true
              idleConnectionTimeoutMinutes: 4
```

**Comment by pbochynski** [comment]
I was unaware we need to split the networks for each availability zone. Thanks for pointing it out. @anishj0shi is the mask /22 ok from your perspective?

**Comment by ebensom** [comment]
@anishj0shi you are right from the perspective of maximum 80 nodes spread over 3 AZs, however for AWS for each each AZ, you need to specify 3 CIDR ranges from the which are subnets of the `nodes` CIDR (one for the actual VM NICs and two for the public and internal ELB backend pools which require one IP for each ELB for each zone), hence comes the requirement to be able to split the specified `nodes` CIDR into (at least) 9. This means that you will need to use at least 4 digits bigger subnet mask for the specific CIDRs when the `nodes` CIDR is split (3 digits bigger would only allow split by 8). 

Having said this, when going with /22 `nodes` CIDR allows 16x /26 subnets (with size of 62 IPs which is sufficient).

When going with /23 `nodes` allows 16x /27 subnets with size 30 IPs each. In this case I would rather go with assigning 3x /26 subnets to workers, while there are still 10x /27 subnets remaining, 6x of which would be used to assign to public and internal ELB subnets. 

When going with /24 `nodes` allows assigning 3x /26 subnet for the 3 worker subnets, while there are still 8x /28 subnets, 6x of which would be used to assign to public and internal ELB subnets. 

Even if we don't have runtimes now which could have more than 80 nodes, I would still be future proof with the assigned subnet CIDR for the workers and would still allow considerably more in the worker IP CIDR than 80, as this field is immutable. When working with /26 subnets for the worker subnets allows 60 IPs per AZ, which is 180 workers in total for a 3-AZ runtime, which seems acceptable for me. 

TL;DR: I would not go down below /26 CIDR for workers (VMs), and hence /24 CIDR for `nodes` attribute of the Shoot,, these are the acceptable minimum to me.

**Comment by anishj0shi** [comment]
Thanks @ebensom for the great explanation. which means, we can already trim down the public and internal subnets from /20 to something way smaller. I particularly dont have strong opinion b/w /22, /23 or /24 as far as it's not /16. This is something, we'll anyways parameterize in future. it's just better to be safe with ipv4 exhaustion in the context of BTP.

> Even if we don't have runtimes now which could have more than 80 nodes, I would still be future proof with the assigned subnet CIDR for the workers and would still allow considerably more in the worker IP CIDR than 80, as this field is immutable. When working with /26 subnets for the worker subnets allows 60 IPs per AZ, which is 180 workers in total for a 3-AZ runtime, which seems acceptable for me.

I agree, but this should be a configurable behavior, based on specific requirements. I'm not even sure, if we should promote creation of Enormous clusters, as the blast radius of production grade application is significantly increased in a single cluster environment. however this is a choice, I'd like to give to the consumer, if they want a one giant cluster or several small clusters privately peered.

**Comment by piotrmiskiewicz** [comment]
@anishj0shi @ebensom  - is it ok that for all providers (Azure, AWS, GCP) I will implement a default `10.250.0.0/22` ?

**Comment by piotrmiskiewicz** [comment]
All IP ranges (nodes, pods ands services) must not overlap. If the Provisioner sends only some of them (for example nodes and pods) the rest will be filled by Gardener with default values. If the provided Pods CIDR overlaps the default services - Gardener will fail:
```
Warning  SchedulingFailed  68s (x17 over 6m37s)  shoot-scheduler  Failed to schedule shoot 'c-46c77da': shoots.core.gardener.cloud "c-46c77da" is forbidden: [spec.networking.services: Invalid value: "100.64.0.0/17": shoot pod network intersects with shoot service network]
```

The solution:

1. The JSON schema contains default values for all CIDRs: nodes, pods and services.
2. If the `networking` field is provided, all 3 CIDRs must be set.


In the UI user see the default values and comments that IP ranges must not overlap.
If the user uses btp cli, all networking fields must be provided.

**Comment by piotrmiskiewicz** [comment]
PR: https://github.com/kyma-project/control-plane/pull/3020

**Comment by piotrmiskiewicz** [comment]
Docs about seed ranges: https://pages.github.tools.sap/kubernetes/gardener/docs/faq/sap-internal/seed-cidr-ranges/

**Comment by piotrmiskiewicz** [comment]
SKR test: https://github.com/kyma-project/kyma-environment-broker/pull/62/files
